### PR TITLE
Tokenizer

### PR DESCRIPTION
**Pull request by ayulockin**
This PR adds the following functionalities:

- train our own tokenizer using HF's Tokenizer library
- batched tokenisation in the `collate_fn`
- correct decoding

Refer readme to see how to run it. 

The `get_tokenizers` function will return two tokenizers -- one for each language.

cc: @ariG23498

### CODE REVIEW

File: src/llm/tokenizer.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        raise FileNotFoundError(
+            f"Tokenizers for {dataset_name} not found. Please train the tokenizers first using python src/llm/train_tokenizer.py."
+        )
+
+
+def decode_tokens(tokenizer, tokens: Union[list[int], list[list[int]]], batch=False):
+    if batch:
+        return tokenizer.decode_batch(tokens)
+    else:
+        return tokenizer.decode(tokens)
```
Comment (ariG23498) [inline src/llm/tokenizer.py:46]: Can we infer the `batch` flag from the `tokens` parameter?

**Review by ariG23498 — CHANGES_REQUESTED** [review]
Do you think adding tokenizer to the `wmt_dataset.py` file would make sense?

My idea is:
1. Add everything in the `tokenizer.py` file to `wmt_dataset.py` file
2. Rename `wmt_dataset.py` to wmt_data_util.py` (or something similar)

Reason behind this is:
1. Keeping things in a single file makes it easier to navigate
2. We would be able to use this template for other datasets and add a single file for each dataset we want our model to train on.

Let me know what you think. After that I will try training a tokenizer and provide a more in-depth review.

**Comment by ayulockin** [comment]
Agree on the single file. Will move `tokenizer.py` to `wmt_data_utils.py`.

File: src/llm/tokenizer.py (hunk 1 above)
Comment (ayulockin) [reply src/llm/tokenizer.py]: I have marked this as a TODO for now. Will do it when we get to it. For now it is just a utility that I used to see if the tokenizer is working fine.

**Comment by ayulockin** [comment]
@ariG23498 the PR is ready for review.

```
from llm import get_dataset, get_tokenizers, get_dataloader

# Load the dataset
ds = get_dataset(name="de-en")

# Get tokenizers
# Path is where you have put the trained tokenizers
tokenizers = get_tokenizers("de-en", path="../data")

# Get dataloader
train_dl = get_dataloader(ds["train"], batch_size=2, source_lang="de", target_lang="en", tokenizer_path="../data")
```

To train the tokenizer you can run this command from the root of your repo:

`python src/llm/wmt_data_utils.py`

If you want to experiment: `python src/llm/wmt_data_utils.py -h` but the defaults are good for now and I have tested them to work well.

**Review by ariG23498 — CHANGES_REQUESTED** [review]
Great work so far.

Left some minor nits. After they are taken care of we can train a tokenizer and push to hub if necessary.

File: src/llm/wmt_data_utils.py [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+    models,
+    normalizers,
+    pre_tokenizers,
+    decoders,
+    trainers,
+)
+from datasets import load_dataset
+
+
+def get_dataset(name):
```
[the remark this answers is not in the document]
Comment (ayulockin) [reply src/llm/wmt_data_utils.py:19]: you are right.

File: src/llm/wmt_data_utils.py [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+                        help='Dataset name (e.g., de-en, fr-en, cs-en). Default is `de-en`')
+    parser.add_argument('--vocab-size', type=int, default=30000,
+                        help='Total vocabulary size for both languages. Default is 30000')
+    parser.add_argument('--min-frequency', type=int, default=2,
+                        help='Minimum frequency for a token to be included while merging the vocabularies. Default is 2')
+    parser.add_argument('--batch-size', type=int, default=1000,
+                        help='Batch size for training. Default is 1000')
+    parser.add_argument('--special-tokens', type=str, nargs='+', default=["<s>", "<pad>", "</s>", "<unk>", "<sep>"],
+                        help='Special tokens to be included in the vocabulary. Default is ["<s>", "<pad>", "</s>", "<unk>", "<sep>"]')
+    args = parser.parse_args()
```
[the remark this answers is not in the document]
Comment (ayulockin) [reply src/llm/wmt_data_utils.py:259]: OH WOW

File: src/llm/wmt_data_utils.py [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+                batch = batch["translation"]
+                yield [ex[lang] for ex in batch]
+
+        # Train tokenizers
+        tokenizer.train_from_iterator(
+            batch_iterator(lang), trainer=trainer, length=len(ds)
+        )
+
+        # Save the trained tokenizers
+        tokenizer.save(f"data/{lang}_tokenizer.json")
```
[the remark this answers is not in the document]
Comment (ayulockin) [reply src/llm/wmt_data_utils.py:240]: can we make a team/org/etc there to keep all the asset in one place? Let me know.

File: src/llm/wmt_data_utils.py [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+
+def create_separate_wmt_tokenizers(
+    vocab_size=30000,
+    min_frequency=2,
+    special_tokens=None,
+    batch_size=1000,
+    dataset_name="de-en",
+):
+    """
+    Create and train two separate tokenizers for English and German from the WMT dataset.
```
[the remark this answers is not in the document]
Comment (ayulockin) [reply src/llm/wmt_data_utils.py:176]: thanks for catching it.

File: src/llm/wmt_data_utils.py [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+    min_frequency: int = 2,
+    special_tokens: list[str] = ["<s>", "<pad>", "</s>", "<unk>", "<sep>"],
+    batch_size: int = 1000,
+    dataset_name: str = "de-en",
+):
+    """
+    Create and train two separate tokenizers for Source and Target languages from the WMT dataset.
+
+    Args:
+        vocab_size (int): Total vocabulary size for the tokenizers. If vocab size is 30000, english and german tokenizers will have 15000 tokens each.
```
Comment (ariG23498) [inline src/llm/wmt_data_utils.py:180]: english -> source
german -> target

Maybe we should have different vocal sizes, and the user should take a call here. WDYT?

File: src/llm/wmt_data_utils.py [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+        min_frequency (int): Minimum frequency for a token to be included in the vocabulary.
+        special_tokens (list): List of special tokens to be included in the vocabulary.
+        batch_size (int): Batch size for the iterator.
+        dataset_name (str): The name of the dataset. Valid names are `de-en`, `fr-en`, `cs-en`, `hi-en`, `ru-en`.
+
+    Returns:
+        Tuple[Tokenizer, Tokenizer]: The trained English and German tokenizers.
```
Comment (ariG23498) [inline src/llm/wmt_data_utils.py:187]: english -> source
german -> target

File: src/llm/wmt_data_utils.py [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+    """
+    # Load training dataset
+    print(f"Loading dataset WMT14:{dataset_name}...")
```
Comment (ariG23498) [inline src/llm/wmt_data_utils.py:190]:
```suggestion
    print(f"Loading dataset WMT14: {dataset_name}...")
```

File: src/llm/wmt_data_utils.py (hunk 6 above)
Comment (ariG23498) [reply src/llm/wmt_data_utils.py:180]: Totally okay with the current setup (I was just thinking how would that be)

We can move this to our next iteration as well.

File: src/llm/wmt_data_utils.py [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+
+        # Define BPE trainer
+        trainer = trainers.BpeTrainer(
+            vocab_size=vocab_size // 2,  # vocab size for each language
+            min_frequency=min_frequency,
+            special_tokens=special_tokens,
+            show_progress=True,
+        )
+
+        # Define batch iterator for English and German
```
Comment (ariG23498) [inline src/llm/wmt_data_utils.py:224]: english -> source
german -> target

**Review by ariG23498 — CHANGES_REQUESTED** [review]
Made a few suggestions.

**Review by ariG23498 — APPROVED** [review]
Super cool!

**Pull request merged**